conmon-rs/server/src/rpc.rs

```diff
@@ -121,7 +121,13 @@ impl conmon::Server for Server {
                     let mut resp = results.get().init_response();
                     // register grandchild with server
                     let child = Child::new(id, grandchild_pid, child.exit_paths);
-                    let exit_tx = capnp_err!(child_reaper.watch_grandchild(child))?;
+
+                    let mut exit_tx = capnp_err!(child_reaper.watch_grandchild(child))?;
+
+                    // Immediately attempt to pull the exit code
+                    // this is to work around a limitation of oneshot that causes an error
+                    // if the reader reads after the sender is closed (goes out of scope).
+                    let ec = capnp_err!(exit_tx.recv().await)?;
```

Comment:
The timeout does not work anymore with this change, let's fix that by using the tokio timeout:
```diff
diff --git a/conmon-rs/server/src/rpc.rs b/conmon-rs/server/src/rpc.rs
index 3e90d9e2..a83e7366 100644
--- a/conmon-rs/server/src/rpc.rs
+++ b/conmon-rs/server/src/rpc.rs
@@ -91,12 +91,12 @@ impl conmon::Server for Server {
     ) -> Promise<(), capnp::Error> {
         let req = pry!(pry!(params.get()).get_request());
         let id = pry!(req.get_id()).to_string();
-        let timeout = req.get_timeout_sec();
+        let timeout_sec = req.get_timeout_sec();
         let pidfile = pry_err!(Console::temp_file_name("exec_sync", "pid"));
 
         debug!(
-            "Got exec sync container request for id {} with timeout {}",
-            id, timeout,
+            "Got exec sync container request for id {} with timeout of {} sec",
+            id, timeout_sec,
         );
 
         let console = pry_err!(Console::new());
@@ -127,14 +127,15 @@ impl conmon::Server for Server {
                     // Immediately attempt to pull the exit code
                     // this is to work around a limitation of oneshot that causes an error
                     // if the reader reads after the sender is closed (goes out of scope).
-                    let ec = capnp_err!(exit_tx.recv().await)?;
+                    let timeout = Duration::from_secs(timeout_sec);
+                    let ec = capnp_err!(capnp_err!(
+                        tokio::time::timeout(timeout, exit_tx.recv()).await
+                    )?)?;
 
                     let mut stdio = vec![];
                     loop {
-                        let msg = if timeout > 0 {
-                            capnp_err!(console
-                                .message_rx()
-                                .recv_timeout(Duration::from_secs(timeout)))?
+                        let msg = if timeout_sec > 0 {
+                            capnp_err!(console.message_rx().recv_timeout(timeout))?
                         } else {
                             capnp_err!(console.message_rx().recv())?
                         };
```

Reply:
ah we can undo the changes done previously, that was a workaround for the "sender closed" issue